File: backend/app/services/prompt_router_service.py
import importlib
import json
import logging
import sys
from pathlib import Path

from alchemy_glyph_router.env_config import resolve_glyph_router_llm_config

logger = logging.getLogger("forge.prompt_router")

_RUNNER = None


def run_prompt_router(theme: str, *, task_id: str | None = None) -> dict:
    cleaned_theme = str(theme or "").strip()
    entered_module_b = False
    glyph_config = resolve_glyph_router_llm_config()
    config_summary = glyph_config.summary()

    logger.info(
        "prompt_router.request %s",
        _json_log(
            {
                "taskId": task_id,
                "themeText": cleaned_theme,
                "moduleBAttempted": True,
                "llmConfig": config_summary,
            }
        ),
    )

    try:
        runner = _load_runner()
        entered_module_b = True
        payload = runner(cleaned_theme, save_output=False, verbose=False)
        logger.info(
            "prompt_router.success %s",
            _json_log(
                {
                    "taskId": task_id,
                    "themeText": cleaned_theme,
                    "moduleBEntered": entered_module_b,
                    "route": payload.get("route_selected"),
                    "routeReason": payload.get("route_reason"),
                    "fallbackApplied": bool(payload.get("fallback_applied", False)),
                    "template": payload.get("final_template"),
                    "model": payload.get("model"),
                    "routeElapsedMs": payload.get("route_elapsed_ms"),
                    "generationElapsedMs": payload.get("generation_elapsed_ms"),
                    "totalElapsedMs": payload.get("total_elapsed_ms"),
                    "llmConfig": config_summary,
                }
            ),
        )
        return payload
    except Exception as exc:
        failure_reason = _summarize_prompt_router_failure(exc)
        logger.error(
            "prompt_router.failure %s",
            _json_log(
                {
                    "taskId": task_id,
                    "themeText": cleaned_theme,
                    "moduleBEntered": entered_module_b,
                    "enteredLocalFallback": True,
                    "fallbackReason": failure_reason,
                    "exceptionType": type(exc).__name__,
                    "exceptionSummary": _exception_summary(exc),
                    "llmConfig": config_summary,
                }
            ),
        )
        raise RuntimeError(failure_reason) from exc


def build_prompt_fallback(theme_text: str) -> str:
    cleaned = str(theme_text or "").strip()
    if not cleaned:
        return "标准等距2.5D游戏俯视视角，纯黑背景，仅展示一个独立技能特效资产，边界清晰，能量受控，中心主体明确。"
    return (
        "标准等距2.5D游戏俯视视角，纯黑背景，仅展示一个独立技能特效资产。"
        f"{cleaned}"
    )


def _load_runner():
    global _RUNNER
    if _RUNNER is not None:
        return _RUNNER

    router_root = Path(__file__).resolve().parents[2] / "alchemy_glyph_router"
    router_path = str(router_root)
    if router_path not in sys.path:
        sys.path.insert(0, router_path)

    logger.info("loading glyph router from %s", router_root)
    module = importlib.import_module("alchemy_glyph_router_api")
    _RUNNER = getattr(module, "run_alchemy_glyph_router")
    return _RUNNER


def _json_log(payload: dict) -> str:
    return json.dumps(payload, ensure_ascii=False, default=str, separators=(",", ":"))


def _exception_summary(exc: Exception) -> str:
    summary = []
    current = exc
    visited = set()
    while current and id(current) not in visited:
        visited.add(id(current))
        text = str(current).strip()
        if text:
            summary.append(text)
        current = current.__cause__ or current.__context__
    return " <- ".join(summary) if summary else type(exc).__name__


def _summarize_prompt_router_failure(exc: Exception) -> str:
    summary = _exception_summary(exc)
    primary = summary.split(" <- ", 1)[0]
    lowered = summary.lower()

    if "missing_env=" in lowered:
        return f"config_missing_env: {primary}"
    if "route_stage_failed:" in lowered:
        return f"route_stage_failed: {primary}"
    if "generation_stage_failed:" in lowered:
        return f"generation_stage_failed: {primary}"
    if "llm_timeout" in lowered or "timed out" in lowered or "readtimeout" in lowered:
        return f"llm_timeout: {primary}"
    if "llm_http_error" in lowered or "http error" in lowered or "status=" in lowered:
        return f"llm_http_error: {primary}"
    if "llm_request_error" in lowered or "connection aborted" in lowered or "connectionerror" in lowered:
        return f"llm_request_error: {primary}"
    if "llm_response_error" in lowered or "invalid_llm_response" in lowered or "empty_llm_content" in lowered:
        return f"llm_response_error: {primary}"
    if "route_response_not_json" in lowered or "invalid_selected_template" in lowered:
        return f"route_parse_error: {primary}"
    if "missing_full_template_mapping=" in lowered:
        return f"template_mapping_error: {primary}"
    return f"module_b_error: {primary}"

File: backend/src/services/prompt-router-service.js
import fs from 'node:fs';
import path from 'node:path';
import { resourcesRoot, resolveGlyphRouterConfig } from '../config.js';

const cardsDir = path.join(resourcesRoot, 'glyph-router-templates', 'cards');
const fullTemplateDir = path.join(resourcesRoot, 'glyph-router-templates', 'full');
const FORMAL_TEMPLATE_CODES = ['A', 'B1', 'B2', 'C', 'D', 'E', 'F'];
const ALLOWED_TEMPLATE_CODES = new Set(FORMAL_TEMPLATE_CODES);
const FULL_TEMPLATE_MAP = {
  A: 'A_container_full',
  B1: 'B1_fallback_full',
  B2: 'B2_rain_full',
  C: 'C_eruption_full',
  D: 'D_trail_full',
  E: 'E_burst_full',
  F: 'F_manifest_full',
};

export function buildPromptFallback(themeText) {
  const cleaned = String(themeText || '').trim();
  if (!cleaned) {
    return 'Standard isometric 2.5D game skill effect asset, pure black background, one controlled alchemy glyph with clear boundaries and a focused center.';
  }
  return `Standard isometric 2.5D game skill effect asset, pure black background, one controlled alchemy glyph. ${cleaned}`;
}

export async function runPromptRouter(themeText, { taskId = null } = {}) {
  const started = Date.now();
  const cleaned = String(themeText || '').trim();
  if (!cleaned) throw new Error('themeText empty: skipped glyph router');

  const routeStarted = Date.now();
  const route = await routeTheme(cleaned);
  const routeElapsedMs = Date.now() - routeStarted;

  let fallbackApplied = false;
  let finalRouteCode = route.selected_template;
  let finalTemplate = FULL_TEMPLATE_MAP[finalRouteCode];
  if (!finalTemplate && route.fallback_needed && route.fallback_target) {
    finalRouteCode = route.fallback_target;
    finalTemplate = FULL_TEMPLATE_MAP[finalRouteCode];
    fallbackApplied = Boolean(finalTemplate);
  }
  if (!finalTemplate) throw new Error(`missing_full_template_mapping=${finalRouteCode}`);

  const generationStarted = Date.now();
  const templateMarkdown = loadFullTemplate(finalTemplate);
  const prompt = await generatePrompt(cleaned, templateMarkdown);
  const generationElapsedMs = Date.now() - generationStarted;
  const totalElapsedMs = Date.now() - started;
  const config = resolveGlyphRouterConfig();
  return {
    final_prompt: prompt,
    route_selected: route.selected_template,
    route_reason: route.route_reason || `gpt route for task ${taskId || '-'}`,
    fallback_applied: fallbackApplied,
    final_template: finalTemplate,
    model: config.model,
    llm_source_family: config.sourceFamily,
    llm_base_url: config.baseUrl,
    llm_api_key_hint: config.apiKey ? `${config.apiKey.slice(0, 4)}...${config.apiKey.slice(-4)}` : null,
    llm_field_sources: config.fieldSources,
    llm_env_files: config.envFilesLoaded,
    route_elapsed_ms: routeElapsedMs,
    generation_elapsed_ms: generationElapsedMs,
    total_elapsed_ms: totalElapsedMs,
  };
}

async function routeTheme(theme) {
  const cards = loadCards();
  const cardBlock = cards
    .map((card) => `Template code: ${card.code}\nTemplate file: ${card.name}\n${card.content}`)
    .join('\n\n');
  const text = await chatOpenAiCompat(
    [
      {
        role: 'system',
        content:
          'You are a template router for video-prompt generation. Choose the best template code only from A, B1, B2, C, D, E, F. Return one JSON object only, with no markdown.',
      },
      {
        role: 'user',
        content:
          `Theme:\n${theme}\n\n` +
          'Return JSON fields: selected_template, route_reason, fallback_needed, fallback_target.\n' +
          'selected_template must be one of A/B1/B2/C/D/E/F. fallback_needed must be true or false. fallback_target should be null when not needed. route_reason should be concise.\n\n' +
          cardBlock,
      },
    ],
    0
  );
  const data = parseJsonObject(text);
  const selectedTemplate = String(data.selected_template || '').trim().toUpperCase();
  if (!ALLOWED_TEMPLATE_CODES.has(selectedTemplate)) {
    throw new Error(`invalid_selected_template=${selectedTemplate}`);
  }
  let fallbackTarget = data.fallback_target;
  if (fallbackTarget === '' || String(fallbackTarget).toLowerCase() === 'null') fallbackTarget = null;
  if (fallbackTarget != null) {
    fallbackTarget = String(fallbackTarget).trim().toUpperCase();
    if (!ALLOWED_TEMPLATE_CODES.has(fallbackTarget)) fallbackTarget = null;
  }
  return {
    selected_template: selectedTemplate,
    route_reason: String(data.route_reason || '').trim(),
    fallback_needed: toBool(data.fallback_needed),
    fallback_target: fallbackTarget,
  };
}

async function generatePrompt(theme, templateMarkdown) {
  const text = await chatOpenAiCompat(
    [
      {
        role: 'system',
        content:
          'You are a Chinese video-generation prompt writer. Generate one complete final Chinese prompt strictly from the given template. Output only the final prompt text, with no title, no explanation, no bullet list, and no markdown.',
      },
      {
        role: 'user',
        content: `Theme:\n${theme}\n\nFull template:\n${templateMarkdown}`,
      },
    ],
    0.7
  );
  return cleanGeneratedPrompt(text);
}

function loadCards() {
  const entries = fs.readdirSync(cardsDir, { withFileTypes: true });
  const cards = [];
  for (const entry of entries) {
    if (!entry.isFile() || !entry.name.endsWith('.md')) continue;
    const code = entry.name.split('_', 1)[0].toUpperCase();
    if (!ALLOWED_TEMPLATE_CODES.has(code)) continue;
    const filePath = path.join(cardsDir, entry.name);
    cards.push({
      code,
      name: entry.name.replace(/\.md$/i, ''),
      content: fs.readFileSync(filePath, 'utf8').trim(),
    });
  }
  cards.sort((a, b) => FORMAL_TEMPLATE_CODES.indexOf(a.code) - FORMAL_TEMPLATE_CODES.indexOf(b.code));
  if (!cards.length) throw new Error('no_template_cards');
  return cards;
}

function loadFullTemplate(templateStem) {
  const filename = templateStem.endsWith('.md') ? templateStem : `${templateStem}.md`;
  try {
    return fs.readFileSync(path.join(fullTemplateDir, filename), 'utf8').trim();
  } catch {
    throw new Error(`missing_full_template=${templateStem}`);
  }
}

async function chatOpenAiCompat(messages, temperature) {
  const config = resolveGlyphRouterConfig();
  if (config.provider !== 'openai_compat') {
    throw new Error(`unsupported_provider=${config.provider}`);
  }
  if (config.missing.length || !config.baseUrl || !config.apiKey || !config.model) {
    throw new Error(`config_missing_env=${config.missing.join(',') || 'LLM_BASE_URL,LLM_API_KEY,OPENAI_COMPAT_MODEL'}`);
  }

  const url = `${config.baseUrl.replace(/\/+$/, '')}/chat/completions`;
  const totalAttempts = 1 + Math.max(0, Number(config.maxRetries || 0));
  let lastError = null;
  for (let attempt = 1; attempt <= totalAttempts; attempt += 1) {
    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${config.apiKey}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          model: config.model,
          messages,
          temperature,
        }),
        signal: AbortSignal.timeout(Math.max(1, Number(config.timeoutSeconds || 30)) * 1000),
      });
      const rawText = await response.text();
      if (!response.ok) {
        throw new Error(`llm_http_error status=${response.status} body=${rawText.slice(0, 200)}`);
      }
      const body = JSON.parse(rawText);
      const content = extractOpenAiMessageText(body?.choices?.[0]?.message?.content);
      if (!content) throw new Error('empty_llm_content');
      return content;
    } catch (error) {
      lastError = error;
    }
  }
  throw new Error(`llm_request_error: ${lastError?.message || 'unknown'}`);
}

function parseJsonObject(rawText) {
  let cleaned = String(rawText || '').trim();
  if (cleaned.startsWith('```')) {
    cleaned = cleaned
      .split(/\r?\n/)
      .filter((line) => !line.trim().startsWith('```'))
      .join('\n')
      .trim();
  }
  const start = cleaned.indexOf('{');
  const end = cleaned.lastIndexOf('}');
  if (start < 0 || end < start) throw new Error('route_response_not_json');
  return JSON.parse(cleaned.slice(start, end + 1));
}

function cleanGeneratedPrompt(text) {
  let cleaned = String(text || '').trim();
  if (cleaned.startsWith('```')) {
    cleaned = cleaned
      .split(/\r?\n/)
      .filter((line) => !line.trim().startsWith('```'))
      .join('\n')
      .trim();
  }
  return cleaned.replace(/^["']|["']$/g, '').trim();
}

function extractOpenAiMessageText(message) {
  if (typeof message === 'string') return message.trim();
  if (Array.isArray(message)) {
    return message
      .map((item) => {
        if (typeof item === 'string') return item;
        if (item?.type === 'text') return item.text || '';
        return '';
      })
      .join('\n')
      .trim();
  }
  return String(message || '').trim();
}

function toBool(value) {
  if (typeof value === 'boolean') return value;
  if (typeof value === 'string') return value.trim().toLowerCase() === 'true';
  return Boolean(value);
}
